File: frontend/src/lib/risk.ts
import type { AgentPermission, RiskLevel } from '@/types/domain';

export const RISK_RANK: Record<RiskLevel, number> = { low: 0, medium: 1, high: 2, critical: 3 };

const BASE_SCORE: Record<RiskLevel, number> = { low: 15, medium: 40, high: 65, critical: 85 };

/**
 * Indicative risk derived from the permissions an agent requests.
 *
 * This is a UX estimate for the frontend. It is NOT a security control and
 * will be replaced by a server-side assessment.
 */
export function deriveRisk(permissions: AgentPermission[]): { level: RiskLevel; score: number } {
  const granted = permissions.filter((p) => p.level !== 'denied');
  if (granted.length === 0) return { level: 'low', score: 5 };

  const level = granted.reduce<RiskLevel>(
    (highest, p) => (RISK_RANK[p.risk] > RISK_RANK[highest] ? p.risk : highest),
    'low',
  );
  const breadth = Math.min(10, granted.length * 2);
  const unguardedHighRisk = granted.filter((p) => RISK_RANK[p.risk] >= RISK_RANK.high && !p.requiresApproval).length;
  const score = Math.min(100, BASE_SCORE[level] + breadth + unguardedHighRisk * 3);
  return { level, score };
}

File: backend/app/services/risk.py
"""Indicative risk derived from requested permissions.

Mirrors the frontend estimate so both show the same number, but the server
value is the authoritative one stored with the agent. It is a heuristic for
prioritisation, not a security control.
"""

from typing import Any

from app.schemas.enums import RISK_LEVELS, RISK_RANK, CapabilityKey, PermissionLevel, RiskLevel

BASE_SCORE: dict[str, int] = {"low": 15, "medium": 40, "high": 65, "critical": 85}

CAPABILITY_BASE_RISK: dict[str, RiskLevel] = {
    "web_access": "medium",
    "api_access": "high",
    "file_access": "medium",
    "database_access": "high",
    "tool_calling": "low",
    "code_execution": "critical",
    "email_send": "high",
}


def permission_risk(capability: CapabilityKey, level: PermissionLevel) -> RiskLevel:
    base = RISK_RANK[CAPABILITY_BASE_RISK[capability]]
    adjusted = base - 1 if level == "read_only" else base + 1 if level == "allowed" else base
    return RISK_LEVELS[max(0, min(len(RISK_LEVELS) - 1, adjusted))]


def derive_risk(permissions: list[dict[str, Any]]) -> tuple[RiskLevel, int]:
    granted = [p for p in permissions if p.get("level") != "denied"]
    if not granted:
        return "low", 5

    level: RiskLevel = "low"
    for permission in granted:
        risk = permission.get("risk", "low")
        if RISK_RANK[risk] > RISK_RANK[level]:
            level = risk

    breadth = min(10, len(granted) * 2)
    unguarded = sum(
        1
        for p in granted
        if RISK_RANK[p.get("risk", "low")] >= RISK_RANK["high"]
        and not p.get("requiresApproval", False)
    )
    score = min(100, BASE_SCORE[level] + breadth + unguarded * 3)
    return level, score
